### ISSUE TITLE
mpileup read-pair overlap detection not detecting all overlaps?

### ISSUE THREAD

**Issue by jeanlain** [samtools/samtools#1036]
Checked on both samtools 1.9 on macOS and samtools 1.5 on Redhat Linux

Steps:
Decompress and index the attached bam file with `samtools index AV-WXa1591_m_scaffold1.bam`
then 
`samtools mpileup -A -Q1 -x -r scaffold_1:5064-5064 WXa1591_m_scaffold1.bam`

The program detects 40G and 13T at that position, which appears correct.

Now run

`samtools mpileup -A -Q1 -r scaffold_1:5064-5064 WXa1591_m_scaffold1.bam`

The program returns 38G and 12T.

Expected results : **36**G and 12T.

My analysis of this file indicates that both mates in **5** read pairs overlap on position 5064, hence ignoring overlaps should result in coverage dropping from 53 to 48, not 53 to 50.
Position 5064 is just a random one. Other positions show this problem. 
Note: this apparent issue may in fact be my misunderstanding of what the "read-pair overlap detection" is supposed to do.

[AV-WXa1591_m_scaffold1.bam.zip](https://github.com/samtools/samtools/files/3080155/AV-WXa1591_m_scaffold1.bam.zip)

**Comment by tseemann** [comment]
Some relevant discussion here?
https://bioinformatics.stackexchange.com/questions/5427/double-counting-coverage-of-overlapped-read-pairs

**Comment by jeanlain** [comment]
No this doesn't seem relevant. It does not discuss issues with overlap detection in mpileup.

**Comment by valeriuo** [comment]
I haven't looked at your data yet, but it looks similar to #986.

**Comment by valeriuo** [comment]
After looking at the data, it appears not be related to #986. In fact, it works as expected.
Two of the 5 reads pairs overlapping position 5064 are:
```
HWI-ST218:508:C77YEANXX:4:1213:7943:80820       161     scaffold_1      5013    60      100M2I23M       =       5022
HWI-ST218:508:C77YEANXX:4:1213:7943:80820       81      scaffold_1      5022    60      91M2I32M        =       5013

HWI-ST218:508:C77YEANXX:6:1209:11011:36200      145     scaffold_1      4969    60      3S122M  =       4970
HWI-ST218:508:C77YEANXX:6:1209:11011:36200      97      scaffold_1      4970    60      121M4S  =       4969
```
which are **not properly paired** (flags 161, 81, 145, 97). The mpileup overlap detection algorithm only works on [mapped and properly paired reads](https://github.com/samtools/htslib/blob/8ab9acf01b5fc95c2a6c18ace1e0a489d0089261/sam.c#L3022), so it is ignoring these two pairs and this is where those two extra bases came from.

**Comment by jmarshall** [comment]
The code says

```c
// mapped mates and paired reads only
if ( node->b.core.flag&BAM_FMUNMAP || !(node->b.core.flag&BAM_FPROPER_PAIR) ) return 0;
```

(And has done since it was accidentally introduced without review in 65617469ffb8e790280448d2d8d61a984135e39b.)

It's not clear that this isn't just a typo and so ought to be `BAM_FPAIRED` (which would give 36G+12T as the OP expected):
* The comment says _paired_, not _properly paired_
* Properly-paired (for most aligners) implies both ends are mapped, so the test condition as it stands is redundant
* I would expect aligners to mark grossly-overlapping read pairs as **not** properly paired, thus checking `BAM_FPROPER_PAIR` would seem to be disabling overlap correction for those read pairs that need it most!

**Comment by jeanlain** [comment]
I didn't know that read pairs had to be mapped in proper pairs. Why this requirement? Paired reads come from the same template, regardless of how they mapped. And if they do overlap, this further support the fact that these reads come from the same template. Including both reads in the counts does not make biological sense to me.

**Comment by jkbonfield** [comment]
I'm not convinced that aligners wouldn't add a properly paired flag to a short insert.  I think it's justified to label these as properly paired too if the insert size distributions happens to show a significant portion are under 2x read length.

Agreed though on the rest of the comments.  It's unclear what the intention was, but I think it's fair that to say users expect paired to mean any pairing.

**Comment by jmarshall** [comment]
@jkbonfield: Yes — I wasn't meaning to say that an aligner couldn't mark such pairs as properly-paired in certain insert distribution scenarios. Just that there exist scenarios in which an aligner would certainly be justified in not marking them properly-paired!

Edited to add “frequently” :smile:

**Comment by tseemann** [comment]
> And if they do overlap, this further support the fact that these reads come from the same
> template. Including both reads in the counts does not make biological sense to me.*

It depends on what you are using the depth for.   For counting such as in RNA-Seq, you would not want to count it twice. But for consensus in genome assembly, both represent a separate "reading" of the the base, just like other reads are?

**Comment by dpryan79** [comment]
> both represent a separate "reading" of the the base, just like other reads are?

They represent separate technical measurements of the same base, but aren't independent since they arise from the same original single stranded fragment. This was the nicely compelling reason behind modifying the consensus Phred score up/down when PE reads overlap.

**Comment by pd3** [comment]
The overlap detection is a hack on top of the original code. It lowers base qualities of overlapping bases so that they are effectively removed from variant calling. This implies that the overlapping bases are counted twice in raw depth calculation, but are counted correctly when counting only high-quality bases. Fixing this would require more modifications of the mpileup engine.

As for the `BAM_FPROPER_PAIR` vs  `BAM_FPAIR` flags, the code was operating under the philosophy "make changes only if you 100% sure". `BAM_FPAIR` is more general, I don't see a problem with switching to it. I am mildly curious why the aligner wouldn't set `BAM_FPROPER_PAIR`.

**Comment by jeanlain** [comment]
If two mates overlap, it's absolutely sure that bases are covered twice by the same template. Why wouldn't they? That the aligner set some particular flag does not alter this certainty, IMO. 
I suppose that if reads are trimmed (by trimmomatic and such), the first base of a read may align within the region mapped by its mate, which could be seen as an inconsistency.

**Comment by jeanlain** [comment]
> > And if they do overlap, this further support the fact that these reads come from the same
> > template. Including both reads in the counts does not make biological sense to me.*
> 
> It depends on what you are using the depth for. For counting such as in RNA-Seq, you would not want to count it twice. But for consensus in genome assembly, both represent a separate "reading" of the the base, just like other reads are?

In this cas, use the "-x" argument in mpileup. You would want to count bases twice even for mates that are properly paired.

**Comment by jeanlain** [comment]
> The overlap detection is a hack on top of the original code. It lowers base qualities of overlapping bases so that they are effectively removed from variant calling.

Well, not if you set `-Q0`, which is precisely what I've been doing. I have my own way of dealing with sequencing errors downstream. (Or one may want to investigate sequencing errors.) The help should state that  setting `-Q0` also disables overlap read detection.

**Comment by tseemann** [comment]
What happens if there is a SNP in the overlapping section
ie. a disagreement at one position ?

**Comment by valeriuo** [comment]
[sam.c:4310](https://github.com/samtools/htslib/blob/3f140ad0bd64beeed2884b818e7b7a08ef06721d/sam.c#L4310)
```c
if ( a_iref+a->core.pos != b_iref+b->core.pos ) continue;   // only CMATCH positions, don't know what to do with indels
```
Also, `cigar_iref2iseq_next` only returns `BAM_CMATCH` (0) or an error code. So the overlap is not removed.

**Comment by 000CU000** [comment]
> 
> 
> What happens if there is a SNP in the overlapping section
> ie. a disagreement at one position ?

I would be also interested what would happen?

**Comment by pd3** [comment]
The higher quality base is used and the lower-quality base is set to BQ=0.
